### PR TITLE
Add ability to report start of paragraphs in braille

### LINKED ISSUE #16895: Add option to display paragraph indicators in braille
### Is your feature request related to a problem? Please describe.
When reading texts in braille, there's no reliable way to know that a new paragraph starts.
### Describe the solution you'd like
In braille settings, a combo box maybe added with options to show an indicator at the start of each paragraph, or disable this indication.
### Describe alternatives you've considered
An add-on maybe used for this feature.
### Additional context
Paragraphs are important and meaningful units in articles and books, useful to structure and understand them.
Sometimes, several blank cells are displayed at the end of paragraphs, but this is not a consistent way to be sure that a new paragraph will start, since the number of cells is variable and this may correspond to spaces.

### PR DESCRIPTION
**Pull request by nvdaes**
### Link to issue number:
Closes #16895
### Summary of the issue:
NVDA should have an option to indicate paragraphs in braille, so that reading, especially books or long pieces of text, can be more comfortable and meaningful.
### Description of user facing changes
- In the braille settings panel, a new combo box can be used for setting NVDA to show two spaces, a pilcrow or no marker at the start of paragraphs in braille. This requires that the reading by paragraph checkbox is enabled.

### Description of development approach
- A new option for this feature has been added in the braille section of config.conf.
- In config["braille"], a string is assigned to get the startParagraphMarker.
- A getParagraphStartMarker function has been added, so that it can be used in different places of braille.py, and in this way we can also make the PILCROW translatable.
- In the getFormatFieldBraille function of braille.py: when the isAtStart parameter is True, NVDA will check if it's configured to read by paragraph and if it should present the start paragraph indicator, through the getParagraphStartMarker function.
- In the _set_windowEndPos function of braille.py: if the braille wrap word and paragraph start options are enabled, and the start window pos to be set is minor than or equal to the lenghts of paragraph start indicator + 1 (1 corresponding to the text separator), tart pos is set to the start position of the last region if it starts with paragraph start indicator plus text separator. This handles some cases where the paragraph start indicator is not displayed when the braille word wrap is enabled.
- The braille settings subpanel has been updated in the gui.
- A new ParagraphStartMarker class has been added to configFlags.py

### Testing strategy:
Tested manually:
- Reading scrolling foreard and backward, and moving forward and backward through paragraphs, the braille display.
- Moving backward and forward between paragraph with right and left arrows, and control+down and control + up arrow.
- Checking the start position to handle cases where the paragraph start indicator wasn't displayed in braille.
- Tested options added in the braille settings panel.
### Known issues with pull request:
None
### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

## Summary by CodeRabbit

- **New Features**
  - Introduced a configuration option to indicate the start of paragraphs in braille for improved accessibility.
  - Added a feature in the Braille settings panel that allows users to toggle the visibility of paragraph starts.

- **Documentation**
  - Updated user guide to include information on the new Braille settings related to paragraph starts.
  - Enhanced user documentation to reflect the impact of paragraph indicators on reading experience and navigation. 

- **Bug Fixes**
  - Improved logic in braille formatting to better reflect user preferences based on updated configuration settings.

### CODE REVIEW

**Comment by nvdaes** [comment]
@coderabbitai review

**Comment by nvdaes** [comment]
@coderabbitai review

File: user_docs/en/changes.md
Comment (CyrilleB79) [inline user_docs/en/changes.md:12]: * Change the order of phrases in the sentence to improve clarity
* Put the real option name
* Use quoting marks

```suggestion
* When the "Read by paragraph" option is enabled, NVDA can now be configured to indicate the start of paragraphs in braille. (#16895,  @nvdaes)
```

File: user_docs/en/userGuide.md
Comment (CyrilleB79) [inline user_docs/en/userGuide.md:2166]:
```suggestion
##### Show paragraphs start {#BrailleShowParagraphStart}
```

File: source/gui/settingsDialogs.py
Comment (CyrilleB79) [inline source/gui/settingsDialogs.py:4378]: The most recent usage is to avoid using settings panel name in the anchor name, in case options are moved in the future. For example, with the increasing number of braille options, we may imagine that we will have 2 separate (and more specific) braille settings panels.

```suggestion
		self.bindHelpEvent("BrailleShowParagraphStart", self.showParagraphStartCombo)
```

File: user_docs/en/userGuide.md
Comment (CyrilleB79) [inline user_docs/en/userGuide.md:2167]: Could you add a table with the possible options and the default one, as done for any recently introduced feature flag option?

File: user_docs/en/userGuide.md
Comment (CyrilleB79) [inline user_docs/en/userGuide.md:2168]: * remove leading dash
* Link the "Read by paragraph" option

```suggestion
If enabled and "[Read by paragraph](#BrailleSettingsReadByParagraph)" is checked, two spaces will be displayed to indicate the start of a paragraph.
```

Moreover, it would be worth rephrasing this paragraph with a bit more details. But I am not familiar with "Read by paragraph" option and will not be able to test on a braille display before next week, so I won't be able to make a detailed suggestion.

Maybe something like the following (to be adapted and polished):
When "Read by paragraph" option is enabled, we face [issueToBeDescribed, e.g. difficulty to identify paragraph start].
Enabling this option enable to mitigate this issue adding two spaces at the beginning of paragraphs.

File: user_docs/en/userGuide.md
Comment (nvdaes) [reply user_docs/en/userGuide.md:2171]: Done.

**Comment by nvdaes** [comment]
@CyrilleB79 , thanks for your review. I think I've addressed all your suggestions. Let me know if I should apply any other changes.

File: user_docs/en/userGuide.md
Comment (XLTechie) [inline user_docs/en/userGuide.md:2173]: English smoothing.

```suggestion
This can be especially helpful in applications used to read large pieces of text, like structured documents or books.
In such documents, knowing where paragraphs start maybe important to understand the structure of the content, or to set bookmarks or annotations based on paragraph position.
```

File: user_docs/en/userGuide.md
Comment (XLTechie) [inline user_docs/en/userGuide.md:2166]: I think the plural of the word "paragraph" is not correct usage here. I'm not 100% sure of that, but reading it in English it just "doesn't sound right".

```suggestion
##### Show paragraph start {#BrailleShowParagraphStart}
```

**Review by XLTechie — CHANGES_REQUESTED** [review]

**Comment by nvdaes** [comment]
@XLTechie , thanks for your review. I've applied your suggestions. Let me know if I should fix other things.

**Comment by nvdaes** [comment]
@coderabbitai review

**Comment by nvdaes** [comment]
These days I'm testing this in my daily reading. I've observed that, once, when scrolling down the braille display, two spaces haven't been shown at the start of the paragrahp and I needed to scroll up and down. This may be related to a separate issue, since sometimes the first character is not shown and scrolling up and down is needed. I thought that trying to set two spaces as the paragraph start indicator may fix this issue, but it doesn't work.
So I think that I won't made more changes for this PR unless required by reviewers.

**Comment by nvdaes** [comment]
@coderabbitai review

**Comment by nvdaes** [comment]
@coderabbitai full review

File: source/braille.py
Comment (seanbudd) [inline source/braille.py:1765]: Can you explain this comment more? what is the case here

File: source/gui/settingsDialogs.py
Comment (seanbudd) [inline source/gui/settingsDialogs.py:4502]: please add a type for event

File: source/braille.py
Comment (nvdaes) [reply source/braille.py:334]: I would like that the paragraph start indicator is customizable. I didn't implement this to start with a simpler gui. We mayuse two spaces by default, and add an edit box to customize this.

File: source/braille.py
Comment (nvdaes) [reply source/braille.py:1765]: When the braille wrap option is enabled, sometimes spaces indicating the start of paragraphs aren't displayed in braille. I'm reading a book in Thorium Reader, a system to read epub and other formats in browse mode, and in specific parts of the book paragraph are displayed without spaces.

**Comment by seanbudd** [comment]
Rather than a custom field,  this should be one setting with 3 options:
- No paragraph start marker (default)
- Use double space to indicate paragraph start
- Use pilcrow to indicate paragraph start i.e. `¶`. It seems like liblouis supports this symbol

File: source/braille.py
Comment (seanbudd) [reply source/braille.py:1880]: can you please expand this comment to explain this case further

File: source/config/configSpec.py
Comment (seanbudd) [inline source/config/configSpec.py:89]:
```suggestion
```

File: source/config/configFlags.py
Comment (seanbudd) [inline source/config/configFlags.py:265]:
```suggestion
			self.NONE: _("No paragraph start marker (default)"),
```

File: source/braille.py
Comment (seanbudd) [inline source/braille.py:1880]:
```suggestion
		# When word wrap is enabled, the first block of spaces may be removed from the current window.
```

File: source/braille.py
Comment (seanbudd) [inline source/braille.py:1882]:
```suggestion
		# This may prevent displaying the start of paragraphs.
```

File: source/gui/settingsDialogs.py
Comment (seanbudd) [inline source/gui/settingsDialogs.py:4394]:
```suggestion
		labelText = _("Paragraph start marker:")
```

File: user_docs/en/userGuide.md
Comment (seanbudd) [inline user_docs/en/userGuide.md:2175]:
```suggestion
In such documents, knowing where paragraphs start may be useful to understand the structure of the content, or to set bookmarks or annotations based on paragraph position.
```

Comment (seanbudd) [inline user_docs/en/userGuide.md:2175]: >or to set bookmarks or annotations based on paragraph position.

Is this through the text editor UX, or are you referring to add-on functionality?

File: user_docs/en/userGuide.md
Comment (seanbudd) [inline user_docs/en/userGuide.md:2170]:
```suggestion
|Options |No paragraph start marker, Double space, Pilcrow|
```

File: user_docs/en/userGuide.md
Comment (seanbudd) [inline user_docs/en/userGuide.md:2171]:
```suggestion
|Default |No paragraph start marker|
```

File: user_docs/en/userGuide.md
Comment (seanbudd) [inline user_docs/en/userGuide.md:2168]: please move this table to the end of the settings group, as is the new standard described in userGuideStandards

File: user_docs/en/userGuide.md
Comment (seanbudd) [inline user_docs/en/userGuide.md:2166]:
```suggestion
##### Paragraph start marker {#BrailleParagraphStartMarkers}
```

File: user_docs/en/userGuide.md
Comment (nvdaes) [reply user_docs/en/userGuide.md:2175]: This is related to applications to read books, for example,Thorium Reader. We can press NVDA+enter to perform a click (activate the current object), sothat the app knows the current element of the DOM (for the current HTML document of the epub being read), and then press control+b toset a bookmark. Then pressing control+n we can open the navigation menu, use the arrow keys to select the bookmarks tab, and create an annotation pressing the edit button, adding several lines of text. If we do this in the middle of a paragraph, the bookmark isadded based on the HTML element, so it won't be very precise. The placeMarkers add-on (maintained by me) considers the number of characters from the start of the text, but bookmarks are added for each document with a different title, so in Thorium I prefer to use the reading app feature, since chapters of books areplaced in separate HTML documents, but the title is for the whole book.

File: user_docs/en/userGuide.md
Comment (nvdaes) [reply user_docs/en/userGuide.md]: Done

**Comment by nvdaes** [comment]
@seanbudd , I think I've addressed all your suggestions. For some reason, a file from include folder has been modified but not in commits made today from me.
AppVeyor has failed due to time exceeding 60 minutes.

**Comment by seanbudd** [comment]
@nvdaes - you must have committed it by mistake, please reset our checkout masters `include/microsoft-ui-uiautomation`

**Review by seanbudd — APPROVED** [review]
Approved, pending include/microsoft-ui-uiautomation getting fixed

**Comment by nvdaes** [comment]
@seanbudd, I've fixed include/microsoft-ui-uiautomation and also a speech file not related to this PR.

File: source/speech/commands.py
Comment (seanbudd) [inline source/speech/commands.py:83]: @nvdaes - this will be added until #16866 is merged

**Review by seanbudd — APPROVED** [review]

File: source/config/configFlags.py
Comment (seanbudd) [inline source/config/configFlags.py:268]: This might help translators understand the symbol better and know how to translate it.

```suggestion
			# Translators: This is a label for a paragraph start marker.
			# Pilcrow is a symbol also known as "paragraph symbol" or "paragraph marker".
			# Ensure this is consistent with other strings with the context "paragraphMarker".
```

File: source/speech/commands.py
Comment (nvdaes) [reply source/speech/commands.py:83]: OK, then I'll revert e1d6241

File: source/config/configFlags.py
Comment (seanbudd) [inline source/config/configFlags.py:259]: Can this also be made translatable?

```suggestion
	# Translators: This is a paragraph start marker used in braille.
	# The default symbol is the pilcrow,
	# a symbol also known as "paragraph symbol" or "paragraph marker".
	# This symbol should translate in braille via LibLouis automatically.
	# If there is a more appropriate character for your locale,
	# consider overwriting this (e.g. the reversed pilcrow ⁋).
	# You can also use Unicode Braille such as ⠘⠏.
	# Ensure this is consistent with other strings with the context "paragraphMarker".
	PILCROW = pgettext("paragraphMarker", "¶")
```

File: user_docs/en/userGuide.md
Comment (seanbudd) [inline user_docs/en/userGuide.md:2174]:
```suggestion
|Options |No paragraph start marker, Double space (  ), Pilcrow (¶)|
```

File: source/config/configFlags.py
Comment (seanbudd) [inline source/config/configFlags.py:269]:
```suggestion
			self.PILCROW: pgettext("paragraphMarker", "Pilcrow (¶)"),
```

File: source/config/configFlags.py
Comment (seanbudd) [inline source/config/configFlags.py:267]:
```suggestion
			self.SPACE: pgettext("paragraphMarker", "Double space (  )"),
```

**Comment by nvdaes** [comment]
OK, pre commit bot fixed e1d6241. Then I won't change anything. Thanks @seanbudd for approving this.

File: source/config/configFlags.py
Comment (seanbudd) [inline source/config/configFlags.py:265]:
```suggestion
			self.NONE: pgettext("paragraphMarker", "No paragraph start marker (default)"),
```

**Review by Qchristensen — APPROVED** [review]
User guide reads fine.  Is it worth clarifying that the pilcrow character is (what some people may know as) a paragraph marker? I must admit I had encountered the character before but never knew its name myself until now

File: source/config/configSpec.py
Comment (seanbudd) [inline source/config/configSpec.py:88]: can you please make this an int to correspond to the index of the option?

File: user_docs/en/userGuide.md
Comment (seanbudd) [inline user_docs/en/userGuide.md:2170]:
```suggestion
In such documents, knowing where paragraphs start may be useful to understand the structure of the content, or to set bookmarks or annotations based on paragraph position.

The options include using double space for softer paragraph breaks, or alternatively the paragraph symbol, Pilcrow (¶), for more obvious paragraph breaks.
```

File: source/config/configFlags.py
Comment (seanbudd) [inline source/config/configFlags.py:264]: Here's a real world example of why this is useful - since several languages have their own symbol for this

```suggestion
	# consider overwriting this (e.g. for Ge'ez ፨).
```

https://en.wikipedia.org/wiki/Pilcrow#Paragraph_signs_in_non-Latin_writing_systems

File: source/config/configSpec.py
Comment (nvdaes) [reply source/config/configSpec.py:88]: Can you provide more details about this, please?

**Comment by nvdaes** [comment]
@seanbudd , many thanks for explaining the reason to make the option about paragraph markers an integer. I didn't realise that s.making pilcrows translatable makes this necessary.
I'll finish and test this properly today, in a few hours.

**Comment by nvdaes** [comment]
@seanbudd , seems that we can not assign a value to PILCROW in the enum class using pgettext or _(). So I've created a new function in braille.py to get the text corresponding to the paragraphStartMarker, and from this functon this can be made translatable.

File: source/config/configFlags.py
Comment (seanbudd) [inline source/config/configFlags.py:266]: this comment should be deleted now that it has been moved to getParagraphStartMarker

File: source/config/configSpec.py
Comment (seanbudd) [inline source/config/configSpec.py:88]: this change is no longer necessary if the pilcrow is not translatable

File: source/braille.py
Comment (seanbudd) [inline source/braille.py:1202]:
```suggestion
def getParagraphStartMarker() -> str | None:
```

**Comment by nvdaes** [comment]
pre-commit.ci run

**Comment by nvdaes** [comment]
@seanbudd , I've addressed your comments. Please let me know if anything else needs to be done.

**Pull request merged**
